### DISCUSSION TITLE
Home/End keys missing functionality in text fields

### DISCUSSION THREAD

**Discussion by smileyhead** [ppy/osu#24555] [Ideas]
Currently, the <kbd>Home</kbd> and <kbd>End</kbd> keys either scroll the page (beatmap listing) or do nothing (seemingly everywhere else), regardless of whether or not a text field has focus <kbd>Shift</kbd> + <kbd>Home</kbd>/<kbd>End</kbd> work as expected). Because of this, quickly moving the cursor to the beginning or end of the field is currently impossible. Could you please implement this?

**Comment by bdach** [comment]
It's not true that <kbd>Home</kbd> or <kbd>End</kbd> _never_ work. Try multiplayer room create textboxes, for instance.

<kbd>Home</kbd> or <kbd>End</kbd> don't work on textboxes which eagerly keep focus (song select, settings menu, beatmap listing) and there are surrounding scrolls, and this was explicitly disabled in https://github.com/ppy/osu/pull/7403. I don't know what the solution is to that conflict.

**Reply by smileyhead** [reply]
The song select conflict specifically happens based on cursor position:

- If the cursor is over the beatmap list, <kbd>Home</kbd> and <kbd>End</kbd> scroll the list.
- If the cursor is over the leaderboard, they scroll that.
- If the cursor is anywhere else (like the space in middle), they do nothing.

Basing keyboard navigation on cursor position sounds odd to me in the first place, but the keys could at least influence the text field when they have nothing else to do with the cursor in the middle.

For the beatmap listing / settings menu, I admit I don't have a clean solution in mind either[^1], but I had an idea: What if, again, the keys influenced the text field if they can't scroll anyway? For example:

- If you're at the top of the results, pressing <kbd>Home</kbd> wouldn't be able to scroll anywhere, so it should instead move the cursor, and vice versa with <kbd>End</kbd>.
- If there's less than one screen's worth of results and so you cannot scroll anyway, both keys only operate the text cursor.

[^1]: Except for adding a toggle setting to choose whether <kbd>Home</kbd> and <kbd>End</kbd> should scroll the page or move the text cursor.
